Add spell checker project

40thieves opened the pull request:
## Changelist

Adds a spell checker project.

I expect this would be a solo project - at least as it stands - but welcome feedback on extensions etc. I initially thought this might be a bit bigger in scope, but many of the interesting aspects of this project are in asking questions about their approach (e.g. "what might happen if you replaced Basic English with real English?"), and I think those are difficult to write a very precise rubric like I think we're aiming for here. Having said that I still think this is a decent standalone project.

SallyMcGrath commented on Project-Spell-Checker/README.md:
```diff
@@ -0,0 +1,49 @@
+- If the word does not exist in the word list, then it should be highlighted below the input as a spelling mistake
+- It should be obvious to the user that there is a spelling mistake, both with visuals and some explanatory text
+- "Correct" words (i.e. those found in the Basic English word list) that are adjacent to punctuation (which is defined as the following characters: `,.?!":;`) should not be marked as spelling mistakes
+- "Incorrect" words (i.e. those not found in the Basic English word list) that are adjacent to punctuation characters should still be marked as spelling mistakes
```
- The website scores 100 for Accessibility on Lighthouse.

SallyMcGrath reviewed (changes requested):
Nice! Simple

It doesn't ask for tests. I think we probably want something, even just console.asserts?

40thieves commented:
Not sure how I missed the testing & lighthouse points, but have now copied them over. Agree that they are both requirements that we want.

illicitonion reviewed (commented):
I just did this project, it's cute, but assuming we're not requiring re-styling input to show invalid words, it feels a little simple? My implementation is ~20 lines of ~trivial code.

If we _are_ requiring re-styling input... That's an interesting, but quite _big_ challenge...

illicitonion commented:
>> If we are requiring re-styling input... That's an interesting, but quite big challenge...
>
> I think this is probably over the line of complexity - it's a surprisingly difficult task, often even for quite senior folks!

Agreed!

> I think we touched on this on a call and my memory is that it sounded like you had some ideas about ramping up the difficulty a bit? (Or possibly could have been Sally?)

I think off the top of my head possible extensions were...
* Adding custom words to the dictionary (and re-evaluating the check)
* Suggesting possible replacement words (possibly with some scaffolding, e.g. we could provide a `calculateLevenshteinDistance` function or something and just expect the trainees to use it)
* Handling punctuation more interestingly (e.g. in-word punctuation, maybe handling hyphenated words as two independent words, passing "blue-green" because both "blue" and "green" are words)
* Special-casing capitalised words assuming they're names ("I went to lunch with Ali" passes even though "ali" isn't in the dictionary)

They're all reasonably minor additions/tweaks, but maybe all of them together equal a reasonable individual project?

40thieves commented:
Pushed a few commits - I've added all of the suggestions except the replacement words. I feel like this might be enough?

Providing `calculateLevenshteinDistance` is interesting but I think this really only adds the challenge of some DOM manipulation?

illicitonion commented on Project-Spell-Checker/README.md:
```diff
@@ -0,0 +1,65 @@
+
+If there are any words that aren't in the Basic English word list, you should consider those words as a spelling mistake. All spelling mistakes should be presented to the user underneath the input, along with some text explaining that the words are misspelled. They should be highlighted in some way that makes it obvious that there is a spelling mistake.
+
+If a misspelled word is detected, there should be a way for the user to add that word to the dictionary. The spell check should then be triggered again, and the word that was added to the dictionary should no longer be considered a misspelled word (so should _not_ be highlighted). **Note**: the custom dictionary does _not_ need to be saved, so reloading the page will reset back to the initial dic ...
```
```suggestion
If a misspelled word is detected, there should be a way for the user to add that word to the dictionary. The spell check should then be triggered again, and the word that was added to the dictionary should no longer be considered a misspelled word (so should _not_ be highlighted). **Note**: the custom dictionary does _not_ need to be saved, so reloading the page can reset back to the initial dictionary.
```

illicitonion reviewed (approved):
Looks great! :shipit:

40thieves commented on Project-Spell-Checker/README.md:
```diff
@@ -0,0 +1,53 @@
+
+Your website should display an input, which allows the user to write some text. Your website should also have some way of triggering a spell check on the text that the user has written.
+
+When the spell check is triggered, your website should check every word that the user has typed to see if it exists in the Basic English word list that we have provided (in `words.json`).
```
This is something of a good point, but... other than the Days Calendar project, I don't see another example of defining how helper files should be imported. Days Calendar is a bit of an exception too as it's importing across web and Node.

So I guess my question is whether we should go through and add instructions throughout all the projects.

40thieves replied:
Actually... now that I think about it, this may fall under the same situation as Days Calendar as (assuming ESM) it'll need a import type assertion.

I need to drop off now, but will try to get to this tomorrow.

illicitonion replied:
Yeah, I think either we want to provide a `.js` file which contains a function to get the data (which can just be trivially `<script>` included), or an example page that does something like output the count of words or something :)

(Happy to remove these bits of scaffolding when we teach more ESM in ITP, but right now we're pretty light on it...)

40thieves replied:
I think I'll look to merge this and follow up with changes for all project in a separate PR.

40thieves commented on Project-Spell-Checker/README.md:
```diff
@@ -0,0 +1,53 @@
+
+When the spell check is triggered, your website should check every word that the user has typed to see if it exists in the Basic English word list that we have provided (in `words.json`).
+
+If there are any words that aren't in the Basic English word list, you should consider those words as a spelling mistake. All spelling mistakes should be presented to the user underneath the input, along with some text explaining that the words are misspelled. They should be highlighted in some way that makes it obvious that there is a spelling mistake.
```
Hmm, good question - I think I was thinking that the text underneath would be sufficient. I think I added the point about highlighting on the idea that the text underneath would be styled to indicate that it was an error (i.e. red, possibly underlining). As discussed below, I think highlighting the word in the input would be much more difficult.

And I will admit "highlighting" is a useful verb to shorthand "mark as misspelled".

40thieves replied:
Pushed a commit to try improve this.

40thieves commented on Project-Spell-Checker/README.md:
```diff
@@ -0,0 +1,53 @@
+- The website has a way for the user to triggering a spell check
+- When triggered, the spell check should check every word in the input to see if it exists in the Basic English word list
+- If the word does not exist in the word list, then it should be highlighted below the input as a spelling mistake
+- It should be obvious to the user that there is a spelling mistake, both with visuals and some explanatory text
```
Good point. Pushed a commit which suggests leaving it blank (I'm thinking about the UX of a "real" spell checker - which usually doesn't explicitly say anything if there's no misspellings)

40thieves commented:
I realised tonight there were more comments I'd missed previously so have pushed a couple more commits to address those.

There's 1 outstanding point about importing a JSON file (plus a general question about importing in these projects), but I'll have to get to that tomorrow.

40thieves commented:
@SallyMcGrath I don't suppose you have a chance to review this today? I'm strongly tempted to (ab)use my admin privs here, since I think I've addressed your points.

40thieves commented:
Nvm, it was blocking the review only because I had an outstanding comment to address. Will merge now.
